test/jdk/com/sun/jndi/dns/ConfigTests/TimeoutWithEmptyDatagrams.java

```diff
@@ -0,0 +1,196 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import jdk.test.lib.net.URIBuilder;
+
+import javax.naming.CommunicationException;
+import javax.naming.Context;
+import javax.naming.directory.InitialDirContext;
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.SocketTimeoutException;
+import java.time.Duration;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+/*
+ * @test
+ * @bug 8339538
+ * @summary Tests that DnsClient correctly calculates left timeout in
+ *          presence of empty datagram packets.
+ * @library ../lib /test/lib
+ * @modules java.base/sun.security.util
+ * @run main/othervm TimeoutWithEmptyDatagrams
+ */
+
+public class TimeoutWithEmptyDatagrams extends DNSTestBase {
+    // initial timeout = 1/4 sec
+    private static final int TIMEOUT = 250;
+    // try 5 times per server
+    private static final int RETRIES = 5;
+    // DnsClient retries again with increased timeout if left
+    // timeout is less than this value, and max retry attempts
+    // is not reached
+    private static final int DNS_CLIENT_MIN_TIMEOUT = 0;
+
+    public TimeoutWithEmptyDatagrams() {
+        setLocalServer(false);
+    }
+
+    public static void main(String[] args) throws Exception {
+        new TimeoutWithEmptyDatagrams().run(args);
+    }
+
+    /*
+     * Tests that we can set the initial UDP timeout interval and the
+     * number of retries.
+     */
+    @Override
+    public void runTest() throws Exception {
+        // Create a DatagramSocket and bind it to the loopback address to simulate
+        // UDP DNS server that doesn't respond
+        try (DatagramSocket ds = new DatagramSocket(new InetSocketAddress(
+                InetAddress.getLoopbackAddress(), 0))) {
+            CountDownLatch gotClientAddress = new CountDownLatch(1);
+            AtomicReference<SocketAddress> clientAddress = new AtomicReference<>();
+            AtomicBoolean stopTestThreads = new AtomicBoolean();
+
+            String allQuietUrl = URIBuilder.newBuilder()
+                    .scheme("dns")
+                    .loopback()
+                    .port(ds.getLocalPort())
+                    .build()
+                    .toString();
+
+            // Run a virtual thread that receives client request packets and extracts
+            // sender address from them.
```

Comment:
I think the lambda could have been encapsulated in a class abstraction to convey clearly its semantics e.g. DnsRequestSink implementing Runnable and started with Thread.startVirtualThread(new DnsRequestSink(. . .));